tests/unit_tests/test_agent_0/test_agent_0.py

```diff
@@ -165,14 +168,71 @@ def check_agent_task(agent_task, *args, **kwargs):
         with self.assertRaises(ValueError):
             agent_0._start_sub_agents()
 
-# ------------------------------------------------------------------------------
 
+    # --------------------------------------------------------------------------
+    #
+    @mock.patch.object(Agent_0, '__init__', return_value=None)
+    def test_start_services(self, mocked_init):
+
+        advanced_services = list()
+
+        def local_advance(services, publish, push):
+
+            nonlocal advanced_services
+            advanced_services = services
+
+        agent_0 = Agent_0()
+        agent_0.advance = local_advance
+        agent_0._log = mock.Mock()
+        agent_0._service_task_ids = list()
 
+        agent_0.services_event = mock.Mock()
+        test_data = [dict({'cores_per_rank': '3', 'executable':'/bin/ls'})]
+        agent_0._cfg = ru.Config(from_dict={'pid':12, 'pilot_sandbox':'/'})
+        agent_0._cfg.services = test_data
+
+        agent_0.services_event.wait = mock.Mock(return_value=True)
+        agent_0._start_services()
+        self.assertTrue(advanced_services[0]['uid'].startswith('service.'))
+        self.assertTrue(test_data[0].items() <=
+                        advanced_services[0]['description'].items())
+
+        agent_0.services_event.wait = mock.Mock(return_value=False)
+        with self.assertRaises(RuntimeError):
+            agent_0._start_services()
+
+
+    # --------------------------------------------------------------------------
+    #
+    @mock.patch.object(Agent_0, '__init__', return_value=None)
+    def test_state_cb_of_services(self, mocked_init):
+
+        agent_0 = Agent_0(ru.Config(), None)
+        agent_0._service_task_ids = ['101','102']
+        agent_0._running_services = ['101']
+        agent_0._log = mock.Mock()
+
+        agent_0.services_event = threading.Event()
+
+        topic = 'test_topic'
+        msg   = {'cmd': 'update',
+                 'arg': [{'uid'  : '101',
+                          'state': 'AGENT_EXECUTING'}]}
+
+        agent_0._state_cb_of_services(topic, msg)
+        agent_0._running_services.append('102')
+
+        self.assertTrue(agent_0.services_event.is_set())
```

Comment:
you shouldn't change `agent_0._running_services` on your own, the method should handle that, e.g., you have an empty `_running_services`, then you provide `msg` with 1st task, and then `msg` with the 2nd task

Reply:
yes agreed